sirepo/template/webcon.py

```diff
@@ -412,6 +515,94 @@ def write_epics_values(server_address, fields, values):
     return True
 
 
+def _add_cells_cluster(notebook, data_var, fields, col_limit, x_var, y_var,
+        x_label, y_label, rpt):
+    notebook.add_code_cell('from sirepo.analysis import ml')
+    clusters_var = 'clusters'
+    scale_var = 'scale'
+    cols = [idx for idx, f in enumerate(fields) if f and
+            idx < col_limit]
+    method_params = PKDict(
+        agglomerative=f'{rpt.clusterCount}',
+        dbscan=f'{rpt.clusterDbscanEps}',
+        gmix=f'{rpt.clusterCount}, {rpt.clusterRandomSeed}',
+        kmeans=f'{rpt.clusterCount}, {rpt.clusterRandomSeed}, {rpt.clusterKmeansInit}',
+    )
+    notebook.add_code_cell(
+        [
+            (f'{scale_var} = sirepo.analysis.ml.scale_data({data_var}[:, {cols}], '
+             f'{[rpt.clusterScaleMin, rpt.clusterScaleMax]})'),
+            (f'{clusters_var} = sirepo.analysis.ml.{rpt.clusterMethod}('
+             f"{scale_var}, "
+             f'{method_params[rpt.clusterMethod]})')
+        ]
+    )
+    # can't use add_report because we don't know the result of
+    # compute_clusters
+    notebook.add_code_cell([
+        'pyplot.figure()',
+        f"pyplot.xlabel('{x_label}')",
+        f"pyplot.ylabel('{y_label}')",
+        "pyplot.title('Clusters')",
+        f'for idx in range(int(max({clusters_var})) + 1):',
+        (f'  cl_x = [x for i, x in enumerate({x_var}) if '
+         f'{clusters_var}[i] == idx]'),
+        (f'  cl_y = [y for i, y in enumerate({y_var}) if '
+         f'{clusters_var}[i] == idx]'),
+        "  pyplot.plot(cl_x, cl_y, '.')",
+        'pyplot.show()'
+    ])
+
+
+def _add_cells_fft(notebook, x_var, y_var, x_label):
+    # only 1 curve (for now?)
+    w_var = 'w'
+    y_norm_var = 'y_norm'
+    notebook.add_code_cell(
+        f'{w_var}, {y_norm_var} = sirepo.analysis.get_fft({x_var}, {y_var})'
+    )
+    y_info = [PKDict(
+        y_var=y_norm_var,
+        y_label='',
+        style='line'
+    )]
+    notebook.add_report(PKDict(
+        x_var=w_var,
+        x_label=x_label,
+        y_info=y_info,
+        title='FFT'
+    ))
+
+
+def _add_cells_fit(notebook, x_var, y_var, x_range_inds_var, rpt):
+    curves = ('fit', 'max', 'min')
+    x_fit_var = 'x_fit'
+    v = f'{x_fit_var}, '
+    for c in curves:
+        v = v + f"{_var('y', c)}, "
+    notebook.add_code_cell(
+        (f'{v}'
+         'p_vals, sigma = sirepo.analysis.fit_to_equation('
```

Comment:
I don't think `sirepo.analysis.fit_to_equation` is a good place to put things. It should be inherited from a class. Use #2939.